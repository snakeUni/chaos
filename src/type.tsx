/**
 * reducer, action 必须拥有一个 type 字段
 */
export type ReducerType<T> = (state: T, action: Action) => T

/**
 * action 必须要有 type 字段用去区分
 */
export interface Action {
  type: string
}

export interface UseStoreResult<T> {
  getState: () => T
  dispatch: (action: Action, uuid: string) => Action
  subscriber: (listener: Listener) => () => void
}

export interface UseSelector<T> {
  getState: () => T
  dispatch: (action: Action) => Action
}

export type UseStore<T> = (reducer: ReducerType<T>, initialState: T) => UseStoreResult<T>

export interface Listener {
  uuid: string
  listener: () => any
}

/**
 * provider 组件
 */
export interface ProviderProps<T> {
  value: UseStoreResult<T>
  children?: React.ReactNode
}

export interface Reducers<T> {
  [key: string]: (state: T, action: Action) => any
}
