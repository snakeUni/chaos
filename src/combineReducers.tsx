import * as React from 'react'
import { Reducers, Action } from './type'

/**
 * 用于把所有的 reducer 组装起来，和 redux 中的 combineReducers 类似
 * @param reducers
 */
export default function combineReducers<T>(reducers: Reducers<T>) {
  const keys = Object.keys(reducers)
  if (keys.length === 0) {
    throw new Error('reducer cannot be null')
  }
  return (state: T, action: Action) => {
    let hasChanged = false
    const nextState = {}
    keys.forEach(key => {
      const reducer = reducers[key]
      const previousState = (state as any)[key]
      const nextStateWithKey = reducer(previousState, action)
      ;(nextState as any)[key] = nextStateWithKey
      hasChanged = hasChanged || nextStateWithKey !== previousState
    })
    return hasChanged ? nextState : state
  }
}
