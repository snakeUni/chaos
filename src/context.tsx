import * as React from 'react'
import { UseStoreResult } from './type'

export const storeContext = React.createContext<UseStoreResult<any> | undefined>(undefined)

/**
 * selectorContext 用户获取值并且可以和 dispatch 相应的值 这个可以进行缓存所以真正存储只的地方是 storeContext
 */
export const selectorContext = React.createContext<UseStoreResult<any> | undefined>(undefined)
