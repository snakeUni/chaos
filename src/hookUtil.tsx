import * as React from 'react'

function forceReducer(state: boolean) {
  return !state
}

/**
 * forceUpdate
 */
export function useForceUpdate() {
  return React.useReducer(forceReducer, false)[1]
}

/**
 * 获取当前的 state 的值
 * @param initialState
 */
export function useCurrent<T>(initialState: T): [() => T, (state: T) => void, T] {
  const [state, setState] = React.useState(initialState)
  const stateRef = React.useRef(initialState)

  const get = () => stateRef.current

  const set = (state: T) => {
    setState(state)
    stateRef.current = state
  }

  return [get, set, state]
}

export function isSame(src: any[], target: any[]) {
  let isSame = true
  for (let i = 0; i < src.length; i++) {
    if (target.indexOf(src[i]) === -1) {
      isSame = false
      break
    }
  }
  return isSame
}
