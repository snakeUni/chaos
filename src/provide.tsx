import * as React from 'react'
import { storeContext, selectorContext } from './context'
import { ProviderProps } from './type'

const { useState } = React

export default function Provider<T>({ value, children }: ProviderProps<T>) {
  const { getState, dispatch } = value
  const [api] = useState({ getState, dispatch })

  return (
    <storeContext.Provider value={value}>
      <selectorContext.Provider value={api}>{children}</selectorContext.Provider>
    </storeContext.Provider>
  )
}
