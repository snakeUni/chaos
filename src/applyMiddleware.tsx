import { compose } from './hookUtil'

/**
 * 直接应用 redux 中的 middleware
 * @param middlewares
 */
export default function applyMiddleware(...middlewares: any[]) {
  return (createStore: any) => (...args: any[]) => {
    const store = createStore(...args)
    let dispatch = (...args: any[]) => {}

    const middlewareAPI = {
      getState: store.getState,
      dispatch: (...args: any[]) => dispatch(...args)
    }
    const chain = middlewares.map(middleware => middleware(middlewareAPI))
    dispatch = compose(...chain)(store.dispatch)

    return {
      ...store,
      dispatch
    }
  }
}
