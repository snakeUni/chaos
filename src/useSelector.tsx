/**
 * useSelector 通过 context 注入相应的信息，如果修改的是当前组件，那么只渲染当前的组件，需要把相应组件用到的 props
 * 全部放入到监控函数中，只要其中发生变化就调用 forceUpdate 触发组件的 re-render, 但是不会影响其他组件的 re-render, 否则就会像 react-redux 的 v6版本
 * 引起不必要渲染导致的的性能问题，最终回退到之前的订阅的版本
 */

import * as React from 'react'
import { selectorContext, SelectorContext } from './context'
import { useForceUpdate } from './hookUtil'

const { useEffect, useContext } = React

export default function useSelector<T>(callbalk: (state: T) => any[]): SelectorContext<T> {
  const api = useContext(selectorContext)
  const forceUpdate = useForceUpdate()
  if (!api) {
    throw new Error('expected useSelector to be used in Provider')
  }
  useEffect(() => {
    forceUpdate(false)
  }, callbalk(api.getState()))
  return api as any
}
