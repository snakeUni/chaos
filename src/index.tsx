import Provider from './provide'
import combineReducers from './combineReducers'
import useSelector from './useSelector'
import useStore from './useStore'
import applyMiddleware from './applyMiddleware'

export { Provider, combineReducers, useSelector, useStore, applyMiddleware }
