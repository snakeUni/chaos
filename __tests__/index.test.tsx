import * as React from 'react'
import { render, cleanup, fireEvent } from '@testing-library/react'
import { Provider, combineReducers, useStore, useSelector } from '../src'

describe('react-redux-chaos test', () => {
  afterEach(cleanup)

  function countReducer(state = 0, action: any) {
    switch (action.type) {
      case 'increase': {
        return state + 1
      }
      case 'decrease': {
        return state - 1
      }
      default:
        return state
    }
  }

  function ageReducer(state = 18, action: any) {
    switch (action.type) {
      case 'increaseBy': {
        return state + action.payload
      }
      case 'decreaseBy': {
        return state - action.payload
      }
      default:
        return state
    }
  }

  const initialState = { count: 0, age: 18 }
  function DemoProvider({ children }: any) {
    const store = useStore(
      combineReducers({ count: countReducer, age: ageReducer }) as any,
      initialState
    )
    return <Provider value={store}>{children}</Provider>
  }

  function CountComponent() {
    const { getState, dispatch } = useSelector(() => ['count'])
    const state: any = getState()

    return (
      <div>
        <div>count: {state.count}</div>
        <div>
          <button onClick={() => dispatch({ type: 'increase' })}>increaseCount</button>
          <button onClick={() => dispatch({ type: 'decrease' })}>decreaseCount</button>
        </div>
      </div>
    )
  }

  function AgeComponent() {
    const { getState, dispatch } = useSelector(() => ['age'])
    const state: any = getState()

    return (
      <div>
        <div>age: {state.age}</div>
        <div>
          <button onClick={() => dispatch({ type: 'increaseBy', payload: 10 })}>increaseAge</button>
          <button onClick={() => dispatch({ type: 'decreaseBy', payload: 5 })}>decreaseAge</button>
        </div>
      </div>
    )
  }

  it('snapshot', () => {
    const { container } = render(
      <DemoProvider>
        <CountComponent />
        <AgeComponent />
      </DemoProvider>
    )
    expect(container.innerHTML).toMatchSnapshot()
  })

  it('trigger count button', () => {
    const { getByText } = render(
      <DemoProvider>
        <CountComponent />
        <AgeComponent />
      </DemoProvider>
    )
    expect(getByText('count: 0')).not.toBeNull()
    fireEvent.click(getByText(/increaseCount/i))
    expect(getByText('count: 1')).not.toBeNull()
    fireEvent.click(getByText(/decreaseCount/i))
    expect(getByText('count: 0')).not.toBeNull()
  })

  it('trigger age button', () => {
    const { getByText } = render(
      <DemoProvider>
        <CountComponent />
        <AgeComponent />
      </DemoProvider>
    )
    expect(getByText('age: 18')).not.toBeNull()
    fireEvent.click(getByText(/increaseAge/i))
    expect(getByText('age: 28')).not.toBeNull()
    fireEvent.click(getByText(/decreaseAge/i))
    expect(getByText('age: 23')).not.toBeNull()
  })
})
