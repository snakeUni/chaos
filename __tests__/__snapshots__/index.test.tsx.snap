// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`react-redux-chaos test snapshot 1`] = `"<div><div>count: 0</div><div><button>increaseCount</button><button>decreaseCount</button></div></div><div><div>age: 18</div><div><button>increaseAge</button><button>decreaseAge</button></div></div>"`;
